File: Engine/Android/core/src/test/kotlin/despia/engine/IntentsConformanceTest.kt
package despia.engine

import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory

/**
 * The SHARED INTENT corpus (OpenSource/Conformance/intents/launch.json) through THIS runtime's
 * Intents core — the Kotlin leg of Core/Intents (F17.6), and the platform the capability
 * actually runs on. The TS reference (packages/kernel/test/intents-conformance.test.ts) and the
 * Swift twin (Engine/iOS/Intents.swift) read the SAME file, so the refusals a cross-platform
 * caller sees are identical and the generated <queries> block cannot drift from the code that
 * needs it.
 */
class IntentsConformanceTest {

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir") ?: ".").absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/intents")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/intents not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun doc(): Map<String, Any?> =
        json(File(corpusDir(), "launch.json").readText()).foundationValue as? Map<String, Any?>
            ?: error("launch.json: not a JSON object")

    @Suppress("UNCHECKED_CAST")
    private fun rows(key: String, minimum: Int): List<Map<String, Any?>> {
        val list = doc()[key] as? List<Map<String, Any?>> ?: error("no $key[]")
        assertTrue(list.size >= minimum, "intents/$key is suspiciously small (${list.size})")
        return list
    }

    private fun str(v: Any?): String = v as? String ?: ""
    private fun name(c: Map<String, Any?>): String = str(c["name"]).ifEmpty { "?" }

    @Suppress("UNCHECKED_CAST")
    private fun strings(v: Any?): List<String> = (v as? List<Any?> ?: emptyList()).map { str(it) }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun flagValuesAndOrder() {
        val d = doc()
        val values = d["flagValues"] as Map<String, Any?>
        assertEquals(values.size, Intents.FLAGS.size, "flag count")
        for ((word, value) in values) {
            assertEquals((value as Number).toInt(), Intents.FLAGS[word], "flag $word")
        }
        assertEquals(strings(d["flagOrder"]), Intents.FLAG_ORDER, "flagOrder")
    }

    private fun stringCases(
        key: String, minimum: Int, fn: (Any?) -> Result<String>,
    ): List<DynamicTest> = rows(key, minimum).map { c ->
        DynamicTest.dynamicTest("intents/$key — ${name(c)}") {
            val got = fn(c["raw"])
            assertEquals(c["ok"] == true, got.isSuccess, "ok")
            if (c["ok"] == true) {
                assertEquals(str(c["expect"]), got.getOrThrow())
            } else {
                assertEquals(str(c["error"]), Intents.code(got.exceptionOrNull()!!), "error")
            }
        }
    }

    @TestFactory
    fun action(): List<DynamicTest> = stringCases("action", 8) { Intents.normalizeAction(it) }

    @TestFactory
    fun packageName(): List<DynamicTest> = stringCases("package", 3) { Intents.normalizePackage(it) }

    @TestFactory
    fun data(): List<DynamicTest> = stringCases("data", 6) { Intents.normalizeData(it) }

    @TestFactory
    fun type(): List<DynamicTest> = stringCases("type", 6) { Intents.normalizeType(it) }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun extras(): List<DynamicTest> = rows("extras", 5).map { c ->
        DynamicTest.dynamicTest("intents/extras — ${name(c)}") {
            val got = Intents.normalizeExtras(c["raw"])
            assertEquals(c["ok"] == true, got.isSuccess, "ok")
            if (c["ok"] == true) {
                assertEquals(c["expect"] as Map<String, Any?>, got.getOrThrow())
            } else {
                assertEquals(str(c["error"]), Intents.code(got.exceptionOrNull()!!), "error")
            }
        }
    }

    @TestFactory
    fun flags(): List<DynamicTest> = rows("flags", 6).map { c ->
        DynamicTest.dynamicTest("intents/flags — ${name(c)}") {
            val got = Intents.foldFlags(c["raw"])
            assertEquals(c["ok"] == true, got.isSuccess, "ok")
            if (c["ok"] == true) {
                assertEquals((c["mask"] as Number).toInt(), got.getOrThrow().mask, "mask")
                assertEquals(strings(c["words"]), got.getOrThrow().words, "words")
            } else {
                assertEquals(str(c["error"]), Intents.code(got.exceptionOrNull()!!), "error")
            }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun intent(): List<DynamicTest> = rows("intent", 6).map { c ->
        DynamicTest.dynamicTest("intents/intent — ${name(c)}") {
            val got = Intents.normalize(c["raw"] as Map<String, Any?>)
            assertEquals(c["ok"] == true, got.isSuccess, "ok")
            if (c["ok"] != true) {
                assertEquals(str(c["error"]), Intents.code(got.exceptionOrNull()!!), "error")
                return@dynamicTest
            }
            val want = c["expect"] as Map<String, Any?>
            val spec = got.getOrThrow()
            assertEquals(str(want["action"]), spec.action, "action")
            assertEquals(str(want["package"]), spec.pkg, "package")
            assertEquals(str(want["data"]), spec.data, "data")
            assertEquals(str(want["type"]), spec.type, "type")
            assertEquals(strings(want["categories"]), spec.categories, "categories")
            assertEquals(want["extras"] as Map<String, Any?>, spec.extras, "extras")
            val wantFlags = want["flags"] as Map<String, Any?>
            assertEquals((wantFlags["mask"] as Number).toInt(), spec.flags.mask, "flags.mask")
            assertEquals(strings(wantFlags["words"]), spec.flags.words, "flags.words")
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun queries(): List<DynamicTest> = rows("queries", 5).map { c ->
        DynamicTest.dynamicTest("intents/queries — ${name(c)}") {
            val specs = (c["specs"] as List<Map<String, Any?>>).map { row ->
                Intents.normalize(
                    mapOf(
                        "action" to row["action"], "data" to row["data"],
                        "type" to row["type"], "package" to row["package"],
                    ),
                ).getOrElse { error("corpus spec must normalize: ${row["action"]}") }
            }
            val want = (c["expect"] as List<Map<String, Any?>>).map { row ->
                Intents.Query(
                    str(row["kind"]), str(row["action"]), str(row["scheme"]),
                    str(row["mimeType"]), str(row["name"]),
                )
            }
            assertEquals(want, Intents.queries(specs))
        }
    }
}

File: Web/packages/kernel/test/intents-conformance.test.ts
//
//  intents-conformance.test.ts — the SHARED INTENT corpus
//  (OpenSource/Conformance/intents/launch.json) through the TS core, the REFERENCE leg of
//  Core/Intents (F17.6). The Kotlin twin (:core IntentsConformanceTest) and the Swift twin
//  (Engine/iOS/Intents.swift) read the SAME file, so the refusals a cross-platform caller sees
//  are identical and the generated <queries> block cannot drift from the code that needs it.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  INTENT_FLAGS, INTENT_FLAG_ORDER,
  normalizeIntentAction, normalizeIntentPackage, normalizeIntentData, normalizeIntentType,
  normalizeIntentExtras, foldIntentFlags, normalizeIntent, intentQueries,
  type IntentResult, type IntentSpec, type RawIntent,
} from "../src/intents.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/intents");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("intents corpus not found");
    dir = parent;
  }
}

type StringCase = { name: string; raw: unknown; ok: boolean; expect?: string; error?: string };
type Doc = {
  flagValues: Record<string, number>;
  flagOrder: string[];
  action: StringCase[];
  package: StringCase[];
  data: StringCase[];
  type: StringCase[];
  extras: { name: string; raw: unknown; ok: boolean; expect?: unknown; error?: string }[];
  flags: { name: string; raw: unknown; ok: boolean; mask?: number; words?: string[]; error?: string }[];
  intent: { name: string; raw: RawIntent; ok: boolean; expect?: unknown; error?: string }[];
  queries: {
    name: string;
    specs: { action: string; data: string; type: string; package: string }[];
    expect: unknown[];
  }[];
};

const doc = JSON.parse(readFileSync(join(corpusDir(), "launch.json"), "utf-8")) as Doc;

assert.ok(doc.action.length >= 8, "intents/action corpus is suspiciously small");
assert.ok(doc.intent.length >= 6 && doc.queries.length >= 5, "intents corpus is suspiciously small");

test("intents — the ABI-frozen flag values and their canonical order", () => {
  assert.deepEqual({ ...INTENT_FLAGS }, doc.flagValues);
  assert.deepEqual([...INTENT_FLAG_ORDER], doc.flagOrder);
});

function stringCases(label: string, cases: StringCase[], fn: (raw: unknown) => IntentResult<string>): void {
  for (const c of cases) {
    test(`intents/${label} — ${c.name}`, () => {
      const got = fn(c.raw);
      assert.equal(got.ok, c.ok, got.ok ? "ok" : `ok (${(got as { detail?: string }).detail ?? ""})`);
      if (got.ok) assert.equal(got.value, c.expect);
      else assert.equal(got.error, c.error, "error");
    });
  }
}

stringCases("action", doc.action, normalizeIntentAction);
stringCases("package", doc.package, normalizeIntentPackage);
stringCases("data", doc.data, normalizeIntentData);
stringCases("type", doc.type, normalizeIntentType);

for (const c of doc.extras) {
  test(`intents/extras — ${c.name}`, () => {
    const got = normalizeIntentExtras(c.raw);
    assert.equal(got.ok, c.ok, "ok");
    if (got.ok) assert.deepEqual(got.value, c.expect);
    else assert.equal(got.error, c.error, "error");
  });
}

for (const c of doc.flags) {
  test(`intents/flags — ${c.name}`, () => {
    const got = foldIntentFlags(c.raw);
    assert.equal(got.ok, c.ok, "ok");
    if (got.ok) {
      assert.equal(got.value.mask, c.mask, "mask");
      assert.deepEqual([...got.value.words], c.words);
    } else {
      assert.equal(got.error, c.error, "error");
    }
  });
}

for (const c of doc.intent) {
  test(`intents/intent — ${c.name}`, () => {
    const got = normalizeIntent(c.raw);
    assert.equal(got.ok, c.ok, got.ok ? "ok" : `ok (${(got as { detail?: string }).detail ?? ""})`);
    if (got.ok) assert.deepEqual(JSON.parse(JSON.stringify(got.value)), c.expect);
    else assert.equal(got.error, c.error, "error");
  });
}

for (const c of doc.queries) {
  test(`intents/queries — ${c.name}`, () => {
    const specs: IntentSpec[] = c.specs.map((s) => {
      const built = normalizeIntent({ action: s.action, data: s.data, type: s.type, package: s.package });
      assert.ok(built.ok, `corpus spec must normalize: ${s.action}`);
      if (!built.ok) throw new Error("unreachable");
      return built.value;
    });
    assert.deepEqual(JSON.parse(JSON.stringify(intentQueries(specs))), c.expect);
  });
}
